File: Assets/DisabledStateCache.ts
import GameManager from "GameManager";
import Asset from "./Asset";
import { isEmpty } from "Helpers";

interface Cache {
  isDisabled: boolean;
  useDisabledAnimation: boolean;
  color: [number, number, number, number];
  showDisabledColor: boolean;
}

export default class DisabledStateCache {
    gameManager: GameManager;
  private cache = new Map<string, Cache>();

  constructor(gameManager: GameManager) {
    this.gameManager = gameManager;
  }

  cacheState(asset: Asset): void {
    this.cache.set(asset.id, {
      isDisabled: asset.isDisabled,
      useDisabledAnimation: asset.useDisabledAnimation,
      color: asset.color,
      showDisabledColor: asset.showDisabledColor,
    });
  }

  restore(idsToSkip?: string[]): void {
    for (const [baseId, disabledState] of this.cache) {
          if (idsToSkip?.includes(baseId)) continue;
          const assets = this.gameManager.assetManager.assets.get(baseId);
          if (!isEmpty(assets)) {
            for (const asset of assets) {
              // Restore the color directly first
              asset.color = disabledState.color;
              // Then set disabled state without applying color changes
              asset.setDisabled(disabledState.isDisabled, {
                useDisabledAnimation: disabledState.useDisabledAnimation,
                showDisabledColor: disabledState.showDisabledColor,
              });
            }
          }
        }
        this.cache.clear();
  }
}

File: Assets/DisabledStateCache.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local Map = ____lualib.Map
local __TS__New = ____lualib.__TS__New
local __TS__ArrayIncludes = ____lualib.__TS__ArrayIncludes
local __TS__Iterator = ____lualib.__TS__Iterator
local ____exports = {}
local ____Helpers = require("Helpers")
local isEmpty = ____Helpers.isEmpty
____exports.default = __TS__Class()
local DisabledStateCache = ____exports.default
DisabledStateCache.name = "DisabledStateCache"
function DisabledStateCache.prototype.____constructor(self, gameManager)
    self.cache = __TS__New(Map)
    self.gameManager = gameManager
end
function DisabledStateCache.prototype.cacheState(self, asset)
    self.cache:set(asset.id, {isDisabled = asset.isDisabled, useDisabledAnimation = asset.useDisabledAnimation, color = asset.color, showDisabledColor = asset.showDisabledColor})
end
function DisabledStateCache.prototype.restore(self, idsToSkip)
    for ____, ____value in __TS__Iterator(self.cache) do
        local baseId = ____value[1]
        local disabledState = ____value[2]
        do
            local ____opt_0 = idsToSkip
            if ____opt_0 and __TS__ArrayIncludes(idsToSkip, baseId) then
                goto __continue5
            end
            local assets = self.gameManager.assetManager.assets:get(baseId)
            if not isEmpty(assets) then
                for ____, asset in ipairs(assets) do
                    asset.color = disabledState.color
                    asset:setDisabled(disabledState.isDisabled, {useDisabledAnimation = disabledState.useDisabledAnimation, showDisabledColor = disabledState.showDisabledColor})
                end
            end
        end
        ::__continue5::
    end
    self.cache:clear()
end
return ____exports
